File: lib/go/nats_server.go
/*
 * Copyright 2017 Workiva
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *     http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package frugal

import (
	"bytes"
	"time"

	"git.apache.org/thrift.git/lib/go/thrift"
	"github.com/nats-io/go-nats"
)

const (
	defaultWorkQueueLen = 64
	defaultWatermark    = 5 * time.Second
)

type frameWrapper struct {
	frameBytes []byte
	timestamp  time.Time
	reply      string
}

// FNatsServerBuilder configures and builds NATS server instances.
type FNatsServerBuilder struct {
	conn          *nats.Conn
	processor     FProcessor
	protoFactory  *FProtocolFactory
	subjects      []string
	queue         string
	workerCount   uint
	queueLen      uint
	highWatermark time.Duration
}

// NewFNatsServerBuilder creates a builder which configures and builds NATS
// server instances.
func NewFNatsServerBuilder(conn *nats.Conn, processor FProcessor,
	protoFactory *FProtocolFactory, subjects []string) *FNatsServerBuilder {
	return &FNatsServerBuilder{
		conn:          conn,
		processor:     processor,
		protoFactory:  protoFactory,
		subjects:      subjects,
		workerCount:   1,
		queueLen:      defaultWorkQueueLen,
		highWatermark: defaultWatermark,
	}
}

// WithQueueGroup adds a NATS queue group to receive requests on.
func (f *FNatsServerBuilder) WithQueueGroup(queue string) *FNatsServerBuilder {
	f.queue = queue
	return f
}

// WithWorkerCount controls the number of goroutines used to process requests.
func (f *FNatsServerBuilder) WithWorkerCount(workerCount uint) *FNatsServerBuilder {
	f.workerCount = workerCount
	return f
}

// WithQueueLength controls the length of the work queue used to buffer
// requests.
func (f *FNatsServerBuilder) WithQueueLength(queueLength uint) *FNatsServerBuilder {
	f.queueLen = queueLength
	return f
}

// WithHighWatermark controls the time duration requests wait in queue before
// triggering slow consumer logic.
func (f *FNatsServerBuilder) WithHighWatermark(highWatermark time.Duration) *FNatsServerBuilder {
	f.highWatermark = highWatermark
	return f
}

// Build a new configured NATS FServer.
func (f *FNatsServerBuilder) Build() FServer {
	return &fNatsServer{
		conn:          f.conn,
		processor:     f.processor,
		protoFactory:  f.protoFactory,
		subjects:      f.subjects,
		queue:         f.queue,
		workerCount:   f.workerCount,
		workC:         make(chan *frameWrapper, f.queueLen),
		quit:          make(chan struct{}),
		highWatermark: f.highWatermark,
	}
}

// fNatsServer implements FServer by using NATS as the underlying transport.
// Clients must connect with the transport created by NewNatsFTransport.
type fNatsServer struct {
	conn          *nats.Conn
	processor     FProcessor
	protoFactory  *FProtocolFactory
	subjects      []string
	queue         string
	workerCount   uint
	workC         chan *frameWrapper
	quit          chan struct{}
	highWatermark time.Duration
}

// Serve starts the server.
func (f *fNatsServer) Serve() error {
	subscriptions := []*nats.Subscription{}
	for _, subject := range f.subjects {
		sub, err := f.conn.QueueSubscribe(subject, f.queue, f.handler)
		if err != nil {
			return err
		}
		subscriptions = append(subscriptions, sub)
	}

	for i := uint(0); i < f.workerCount; i++ {
		go f.worker()
	}

	logger().Info("frugal: server running...")
	<-f.quit
	logger().Info("frugal: server stopping...")

	for _, sub := range subscriptions {
		sub.Unsubscribe()
	}

	return nil
}

// Stop the server.
func (f *fNatsServer) Stop() error {
	close(f.quit)
	return nil
}

// handler is invoked when a request is received. The request is placed on the
// work channel which is processed by a worker goroutine.
func (f *fNatsServer) handler(msg *nats.Msg) {
	if msg.Reply == "" {
		logger().Warn("frugal: discarding invalid NATS request (no reply)")
		return
	}
	select {
	case f.workC <- &frameWrapper{frameBytes: msg.Data, timestamp: time.Now(), reply: msg.Reply}:
	case <-f.quit:
		return
	}
}

// worker should be called as a goroutine. It reads requests off the work
// channel and processes them.
func (f *fNatsServer) worker() {
	for {
		select {
		case <-f.quit:
			return
		case frame := <-f.workC:
			dur := time.Since(frame.timestamp)
			if dur > f.highWatermark {
				logger().Warnf("frugal: request spent %+v in the transport buffer, your consumer might be backed up", dur)
			}
			if err := f.processFrame(frame.frameBytes, frame.reply); err != nil {
				logger().Errorf("frugal: error processing request: %s", err.Error())
			}
		}
	}
}

// processFrame invokes the FProcessor and sends the response on the given
// subject.
func (f *fNatsServer) processFrame(frame []byte, reply string) error {
	// Read and process frame.
	input := &thrift.TMemoryBuffer{Buffer: bytes.NewBuffer(frame[4:])} // Discard frame size
	// Only allow 1MB to be buffered.
	output := NewTMemoryOutputBuffer(natsMaxMessageSize)
	iprot := f.protoFactory.GetProtocol(input)
	oprot := f.protoFactory.GetProtocol(output)
	if err := f.processor.Process(iprot, oprot); err != nil {
		return err
	}

	if !output.HasWriteData() {
		return nil
	}

	// Send response.
	return f.conn.Publish(reply, output.Bytes())
}

File: lib/python/frugal/aio/server/nats_server.py
# Copyright 2017 Workiva
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#     http://www.apache.org/licenses/LICENSE-2.0
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import logging
import struct
from typing import List

from nats.aio.client import Client
from thrift.Thrift import TApplicationException
from thrift.transport.TTransport import TMemoryBuffer

from frugal import _NATS_MAX_MESSAGE_SIZE
from frugal.aio.processor import FProcessor
from frugal.protocol import FProtocolFactory
from frugal.server import FServer
from frugal.transport import TMemoryOutputBuffer

logger = logging.getLogger(__name__)


class FNatsServer(FServer):
    """FNatsServer is an FServer that uses nats as the underlying transport."""
    def __init__(
            self,
            nats_client: Client,
            subjects: List[str],
            processor: FProcessor,
            protocol_factory: FProtocolFactory,
            queue=''
    ):
        self._nats_client = nats_client
        self._subjects = [subjects] if isinstance(subjects, str) else subjects
        self._processor = processor
        self._protocol_factory = protocol_factory
        self._queue = queue
        self._sub_ids = []

    async def serve(self):
        """Subscribe to the server subject and queue."""
        self._sub_ids = []
        for subject in self._subjects:
            self._sub_ids.append(await self._nats_client.subscribe_async(
                subject,
                queue=self._queue,
                cb=self._on_message_callback,
            ))
        logger.info('Frugal server running...')

    async def stop(self):
        """Unsubscribe from the server subject."""
        for sid in self._sub_ids:
            await self._nats_client.unsubscribe(sid)

    async def _on_message_callback(self, message):
        """The function to be executed when a message is received."""
        if not message.reply:
            logger.warn('no reply present, discarding message')
            return

        frame_size = struct.unpack('!I', message.data[:4])[0]
        if frame_size > _NATS_MAX_MESSAGE_SIZE - 4:
            logger.warning('frame size too large, dropping message')
            return

        # process frame, first four bytes are the frame size
        iprot = self._protocol_factory.get_protocol(
            TMemoryBuffer(message.data[4:])
        )
        otrans = TMemoryOutputBuffer(_NATS_MAX_MESSAGE_SIZE)
        oprot = self._protocol_factory.get_protocol(otrans)

        try:
            await self._processor.process(iprot, oprot)
        except TApplicationException:
            # Continue so the exception is sent to the client
            pass
        except Exception:
            return

        if len(otrans) == 4:
            return

        await self._nats_client.publish(message.reply, otrans.getvalue())
